src/main/scala/is/hail/stats/SkatModel.scala

```diff
@@ -0,0 +1,220 @@
+package is.hail.stats
+
+import breeze.linalg._
+import breeze.numerics._
+import com.sun.jna.Native
+import com.sun.jna.ptr.IntByReference
+import is.hail.utils._
+
+/** SKATModel
+  *   Computes the p value for the skatStat computed by the  SKAT algorithm.
+  * Input:
+  *   weightedG               (n x k) DenseMatrix
+  *     The genotype matrix associated with n samples over k genetic variants
+  *     each of the variants have been scaled by the weight matrix.
+  *   GwtQ                     (m x k) DenseMatrix
+  *     The weighted genotype matrix rotated by the reduced Q coming from the
+  *     rleast squares problem being solved in the SKAT algorithm.
+  *   skatStat                Double
+  *     The point at which the cumulative distribution needs to be evaluated
+  *     at. This score comes from the norm of the residual in the weighted
+  *     Genotype space.
+  * Returns:
+  *   (skatStat, pValue)      (Double,Double)
+  *     The skatStat and the pValue computed by Davies Algorithm are returned
+  *     as a pair.
+  */
+
+class
+SkatModel(weightedG: DenseMatrix[Double],QtGw: DenseMatrix[Double], skatStat: Double) {
+
+  /** Davies Algorithm original C code
+    *   Citation:
+    *     Davies, Robert B. "The distribution of a linear combination of
+    *     x2 random variables." Applied Statistics 29.3 (1980): 323-333.
+    *   Software Link:
+    *     http://www.robertnz.net/QF.htm
+    */
+  @native
+  def qfWrapper(lb1: Array[Double], nc1: Array[Double], n1: Array[Int], r1: Int,
+    sigma: Double, c1: Double, lim1: Int, acc: Double, trace: Array[Double],
+    ifault: IntByReference): Double
+
+  Native.register("hail")
+
+  def computeSkatStats(): (Double, Double) = {
+
+    //Compute coefficients of chi-squared distribution
+    val variantGrammian = 0.5 * (weightedG.t * weightedG - QtGw.t * QtGw)
+    val allEvals = eigSymD.justEigenvalues(variantGrammian).toArray
+
+    //println("Grammian", variantGrammian)
+
+    //compute threshold for eigenvalues
+    var allEvalsSum = 0.0
+    var i = 0
+    while(i < allEvals.length){
+      allEvalsSum += allEvals(i)
+      i += 1
+    }
+    val threshold = 1e-5 * allEvalsSum / allEvals.length
+
+    //Initialize parameters to Davies' algorithm
+    val evals = allEvals.filter(_ > threshold)
+    val terms = evals.length
+
+    //println("new case")
+    //evals.foreach(println(_))
+
+    val noncentrality = Array.fill[Double](terms)(0.0)
+    val dof = Array.fill[Int](terms)(1)
+    val trace = Array.fill[Double](7)(0.0)
```

Comment:
there are 7 values the algorithm stores in the array which correspond to information about the numerical integration routine. We don't need it once we're done, but the algorithm  requires it. I could always just initialize these arrays inside the qfWrapper function in the C code rather than passing them in if you think that would be better.